crates/wasmtime/src/ref.rs

```diff
@@ -2,224 +2,179 @@
 
 use std::any::Any;
 use std::cell::{self, RefCell};
+use std::convert::TryFrom;
 use std::fmt;
+use std::marker::PhantomData;
 use std::rc::{Rc, Weak};
-
-trait InternalRefBase: Any {
-    fn as_any(&self) -> &dyn Any;
-    fn host_info(&self) -> Option<cell::RefMut<Box<dyn Any>>>;
-    fn set_host_info(&self, info: Option<Box<dyn Any>>);
-    fn ptr_eq(&self, other: &dyn InternalRefBase) -> bool;
-}
-
-#[derive(Clone)]
-pub struct InternalRef(Rc<dyn InternalRefBase>);
-
-impl InternalRef {
-    pub fn is_ref<T: 'static>(&self) -> bool {
-        let r = self.0.as_any();
-        Any::is::<HostRef<T>>(r)
-    }
-    pub fn get_ref<T: 'static>(&self) -> HostRef<T> {
-        let r = self.0.as_any();
-        r.downcast_ref::<HostRef<T>>()
-            .expect("reference is not T type")
-            .clone()
-    }
-}
-
-struct AnyAndHostInfo {
-    any: Box<dyn Any>,
-    host_info: Option<Box<dyn Any>>,
-}
-
-#[derive(Clone)]
-pub struct OtherRef(Rc<RefCell<AnyAndHostInfo>>);
+use wasmtime_runtime::VMExternRef;
 
 /// Represents an opaque reference to any data within WebAssembly.
 #[derive(Clone)]
-pub enum ExternRef {
-    /// A reference to no data.
-    Null,
-    /// A reference to data stored internally in `wasmtime`.
-    Ref(InternalRef),
-    /// A reference to data located outside of `wasmtime`.
-    Other(OtherRef),
+pub struct ExternRef {
+    pub(crate) inner: VMExternRef,
+    pub(crate) store: Weak<crate::runtime::StoreInner>,
```

Comment:
The only reason this is here is for the `host_info` APIs, right? If so I would personally prefer that we just jettison that support from the C API for now. I think we should find a better way for implementing that, if necessary, instead of altering the design of the `wasmtime` crate itself.